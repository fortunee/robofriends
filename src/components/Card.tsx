import React from 'react';

interface CardProps {
  id: number,
  name: string,
  email: string,
}

const Card = ({ id, name, email }: CardProps) => (
  <div className="tc bg-light-blue dib br3 pa3 ma2 grow shadow-5">
    <img src={`https://robohash.org/${id}?size=200x200`} alt="robots"/>
    <div>
      <h2>{name}</h2>
      <p>{email}</p>
    </div>
  </div>
);

export default Card;
