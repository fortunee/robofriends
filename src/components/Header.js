import React from 'react';

const Header = () => <h1 className="f1">Robo friends</h1>;

export default Header;
