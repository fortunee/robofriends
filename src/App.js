import React, { Component } from 'react';
import CardList from './CardList';
import SearchBox from './SearchBox';
import Loading from './Loading';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      robots: [],
      searchField: '',
    };
  }

  componentDidMount() {
    fetch('https://jsonplaceholder.typicode.com/users')
      .then((res) => res.json())
      .then((robots) => this.setState({ robots }));
  }

  onSearchChange = (e) => {
    this.setState({ searchField: e.target.value });
  };

  render() {
    const filteredRobots = this.state.robots.filter((r) =>
      r.name
        .toLocaleLowerCase()
        .includes(this.state.searchField.toLocaleLowerCase())
    );

    if (this.state.robots.length > 0) {
      return (
        <div className="tc">
          <h1 className="f1">Robo friends</h1>
          <SearchBox searchChange={this.onSearchChange} />
          <CardList robots={filteredRobots} />
        </div>
      );
    } else {
      return <Loading />;
    }
  }
}

export default App;
