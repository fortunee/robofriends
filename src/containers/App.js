import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  CardList,
  ErrorBoundary,
  Loading,
  SearchBox,
  Scroll,
} from '../components';

import { requestRobots, setSearchField } from '../actions';

import './App.css';

const App = () => {
  const dispatch = useDispatch();
  const { searchField, robots, isPending } = useSelector((state) => ({
    searchField: state.searchRobots.searchField,
    robots: state.requestRobots.robots,
    isPending: state.requestRobots.isPending,
  }));

  useEffect(() => {
    dispatch(requestRobots());
  }, [dispatch]);

  const onSearchChange = (e) => dispatch(setSearchField(e.target.value));

  const filteredRobots = robots.filter((robot) =>
    robot.name.toLocaleLowerCase().includes(searchField.toLocaleLowerCase())
  );

  return isPending ? (
    <Loading />
  ) : (
    <div className="tc">
      <h1 className="f1">Robo friends</h1>
      <SearchBox searchChange={onSearchChange} />
      <Scroll>
        <ErrorBoundary>
          <CardList robots={filteredRobots} />
        </ErrorBoundary>
      </Scroll>
    </div>
  );
};

export default App;
