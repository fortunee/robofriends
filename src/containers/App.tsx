import React, { ChangeEvent, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  CardList,
  ErrorBoundary,
  Header,
  Loading,
  SearchBox,
  Scroll,
} from '../components';

import { requestRobots, setSearchField } from '../actions';

import './App.css';

export interface RobotProps {
  id: number;
  name: string;
  email: string;
  username?: string;
}

interface State {
  searchRobots: { searchField: string };
  requestRobots: {
    robots: Array<RobotProps>;
    isPending: boolean;
    error: string;
  };
}

const App = () => {
  const dispatch = useDispatch();
  const { searchField, robots, isPending, hasError } = useSelector(
    (state: State) => ({
      searchField: state.searchRobots.searchField,
      robots: state.requestRobots.robots,
      isPending: state.requestRobots.isPending,
      hasError: state.requestRobots.error,
    })
  );

  useEffect(() => {
    dispatch(requestRobots());
  }, [dispatch]);

  const onSearchChange = (e: ChangeEvent<HTMLInputElement>) =>
    dispatch(setSearchField(e.target.value));

  const filteredRobots = robots.filter((robot: RobotProps) =>
    robot.name.toLocaleLowerCase().includes(searchField.toLocaleLowerCase())
  );

  return isPending || hasError ? (
    <Loading />
  ) : (
    <div className="tc">
      <Header />
      <SearchBox searchChange={onSearchChange} />
      <Scroll>
        <ErrorBoundary>
          <CardList robots={filteredRobots} />
        </ErrorBoundary>
      </Scroll>
    </div>
  );
};

export default App;
