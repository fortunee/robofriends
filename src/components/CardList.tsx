import React from 'react';
import { RobotProps } from '../containers/App';
import Card from './Card';

interface CardListProps {
  robots: Array<RobotProps>
}

const CardList = ({ robots }: CardListProps) => (
  <>
    {robots.map((robot) => (
      <Card
        key={robot.id}
        id={robot.id}
        name={robot.name}
        email={robot.email}
      />
    ))}
  </>
);

export default CardList;
