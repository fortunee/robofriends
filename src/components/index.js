import Card from './Card';
import CardList from './CardList';
import Loading from './Loading';
import Scroll from './Scroll';
import SearchBox from './SearchBox';
import ErrorBoundary from './ErrorBoundary';

export { Card, CardList, Loading, Scroll, SearchBox, ErrorBoundary };
