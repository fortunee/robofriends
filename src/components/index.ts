import Card from './Card/Card';
import CardList from './CardList/CardList';
import Header from './Header/Header';
import Loading from './Loading/Loading';
import Scroll from './Scroll/Scroll';
import SearchBox from './SearchBox/SearchBox';
import ErrorBoundary from './ErrorBoundary/ErrorBoundary';

export { Card, CardList, Header, Loading, Scroll, SearchBox, ErrorBoundary };
