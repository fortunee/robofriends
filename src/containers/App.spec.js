import { render, mount } from 'enzyme';
import React from 'react';
import * as redux from 'react-redux';
import { SearchBox } from '../components';

import App from './App';

describe('App', () => {
  const useSelectorMock = jest.spyOn(redux, 'useSelector');
  const useDispatchMock = jest.spyOn(redux, 'useDispatch');

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  it('Should render App snapshot', () => {
    useDispatchMock.mockReturnValue({
      robots: [],
      searchField: '',
    })
    useSelectorMock.mockReturnValue({
      robots: [],
      searchField: '',
    });
    const component = render(<App />);
    expect(component).toMatchSnapshot();
  })

  it('Should render loading component if request is pending', () => {
    useSelectorMock.mockReturnValue({
      robots: [],
      isPending: true,
    });

    const component = render(<App />);
    expect(component.text()).toEqual('Loading...');
  })

  it('Should call dispatch', () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);
    useSelectorMock.mockReturnValue({
      robots: [{
        id: 1,
        name: 'Jane',
        username: 'jdoe',
        email: 'jane@doe'
      }],
      searchField: '',
      isPending: true,
    });

    expect(dummyDispatch).not.toHaveBeenCalled();
    mount(<App />);
    expect(dummyDispatch).toHaveBeenCalled();
  });

  it('should call searchChange', () => {
    const dummyDispatch = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);
    useSelectorMock.mockReturnValue({
      robots: [],
      searchField: ''
    });

    const component = mount(<App />);
    const SearchBoxWrapper = component.find(SearchBox).find('input[type="search"]');
    expect(SearchBoxWrapper.length).toEqual(1)
    const mockedEvent = { target: { value: 'something'} };
    SearchBoxWrapper.simulate('change', mockedEvent);
  })
})