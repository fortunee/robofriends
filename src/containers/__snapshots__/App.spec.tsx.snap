// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render App snapshot 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "tc",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "f1",
        },
        "children": Array [
          Node {
            "data": "Robo friends",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "h1",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "pa2",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "aria-label": "Search Robos",
                "class": "pa3 ba b--green",
                "placeholder": "Search robo friends",
                "type": "search",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-label": undefined,
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-label": undefined,
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Node {
            "attribs": Object {
              "style": "overflow-y:scroll;height:80vh",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "pa2",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "aria-label": "Search Robos",
              "class": "pa3 ba b--green",
              "placeholder": "Search robo friends",
              "type": "search",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-label": undefined,
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-label": undefined,
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "style": "overflow-y:scroll;height:80vh",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "f1",
          },
          "children": Array [
            Node {
              "data": "Robo friends",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "h1",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {
          "style": "overflow-y:scroll;height:80vh",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "pa2",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "aria-label": "Search Robos",
                "class": "pa3 ba b--green",
                "placeholder": "Search robo friends",
                "type": "search",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-label": undefined,
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-label": undefined,
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Node {
            "attribs": Object {
              "class": "f1",
            },
            "children": Array [
              Node {
                "data": "Robo friends",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "h1",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
